### PR TITLE
feat: metrics route

### PR DESCRIPTION
**Pull request by Terkwood**
Exposes a `__metrics` route.  Local testing shows that it produces some values.

```
❯ curl localhost:9000/__metrics
```
...

```
# HELP jvm_buffer_pool_used_bytes Used bytes of a given JVM buffer pool.
# TYPE jvm_buffer_pool_used_bytes gauge
jvm_buffer_pool_used_bytes{pool="direct",} 4.1310316E7
...snip...
```

### CODE REVIEW

File: orchard-ws/app/controllers/MetricController.scala [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+import scala.concurrent.ExecutionContext
+
+import play.api.mvc._
+
+import services.Metric
+
+@Singleton
+class MetricController @Inject() (
+                                   cc: ControllerComponents,
```
Comment (damianxu88) [inline orchard-ws/app/controllers/MetricController.scala:13]: @Terkwood formatting needs a fix

Comment (Terkwood) [reply orchard-ws/app/controllers/MetricController.scala]: fixed this file, thanks

File: orchard-ws/conf/application.conf [hunk 2]
```diff
@@ -359,3 +359,10 @@ play.assets {
   path = "/public"
   urlPrefix = "/assets"
 }
+
+orchard.metrics {
+  enable = true
+
+  # Disable metric collection for these request paths
+  bypass.paths = ["/__status", "/", "/__metrics"]
+}
```
Comment (realstraw) [inline orchard-ws/conf/application.conf:368]: newline

**Review by realstraw — COMMENTED** [review]
@schowsf can you take a look too?

**Comment by damianxu88** [comment]
@Terkwood @realstraw @schowsf am thinking we could directly integrate OSS version of kinetic-pulse into orchard vs the hardcoded implementation here.

**Comment by Terkwood** [comment]
Compare to 

- https://github.com/salesforce/acdc/blob/master/acdc-ws/app/services/Metric.scala
- https://github.com/salesforce/acdc/blob/master/acdc-ws/app/controllers/MetricController.scala
- https://github.com/salesforce/acdc/blob/master/acdc-ws/conf/application.conf
- https://github.com/salesforce/acdc/blob/master/acdc-ws/conf/routes

**Comment by schowsf** [comment]
I think this needs a Play Filter.  reference these in acdc.  

https://github.com/salesforce/acdc/blob/master/acdc-ws/app/utils/MetricFilter.scala
https://github.com/salesforce/acdc/blob/master/acdc-ws/app/Filters.scala

**Comment by Terkwood** [comment]
Oh OK thanks, will add

**Comment by schowsf** [comment]
> @Terkwood @realstraw @schowsf am thinking we could directly integrate OSS version of kinetic-pulse into orchard vs the hardcoded implementation here.

@Terkwood https://github.com/salesforce/kineticpulse is a refactor of the logic from acdc to be a library.  We are in the process of making kineticpulse open source, with 1 or 2 steps still pending. The README there is still mostly relevant.

**Comment by Terkwood** [comment]
@schowsf i'm fiddling with the filter config.  like you said, that code exists in ACDC. but it's not actually used in application.conf, nor is it bound in Module.scala.  might take me a little while to figure out the guice error that i'm getting , which seems to _only impact testing_ -- webservice itself seems to still be basically OK

```
[info] FunctionalSpec:
[info] FunctionalSpec *** ABORTED ***
[info]   com.google.inject.ProvisionException: Unable to provision, see the following errors:
[info]
[info] 1) Error injecting constructor, java.lang.IllegalArgumentException: Failed to register Collector of type Counter: http_requests_total is already in use by another Collector of type Counter
[info]   at services.PrometheusMetric.<init>(Metric.scala:60)
[info]   at services.PrometheusMetric.class(Metric.scala:60)
[info]   while locating services.PrometheusMetric
[info]   while locating services.Metric
[info]     for the 1st parameter of utils.MetricFilter.<init>(MetricFilter.scala:13)
[info]   while locating utils.MetricFilter
[info]     for the 1st parameter of Filters.<init>(Filters.scala:9)
[info]   while locating Filters
[info]   while locating play.api.http.HttpFilters
[info]     for the 6th parameter of play.api.http.JavaCompatibleHttpRequestHandler.<init>(HttpRequestHandler.scala:322)
```

I haven't pushed this code up to the repo yet because it seems like it's a step in the wrong direction until I get the test sorted out.  At this point I just have a small change to application.conf to enable the filter, and a broken Module.scala

**Comment by Terkwood** [comment]
Woo I got this annoying test thing worked out.  Here comes something nice 👯‍♂️

File: orchard-ws/app/services/Metric.scala [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  def startApiTimer(labels: String*): () => Unit
+
+  def collect: Future[String]
+
+  def clear(): Unit
+
+}
+
+@Singleton
+class PrometheusMetric @Inject() (implicit ec: ExecutionContext) extends Metric {
```
Comment (schowsf) [inline orchard-ws/app/services/Metric.scala:60]: @Terkwood good that you worked out the test error.  

Another option to deal with that is to move the metric register to a companion `object PrometheusMetric`.  The Play Singleton did not work as well as the Scala `object`

Comment (Terkwood) [reply orchard-ws/app/services/Metric.scala:60]: Clarified offline with Simon that we can follow this path in a future refactor to use kineticpulse -- marking resolved

**Review by realstraw — APPROVED** [review]

**Pull request merged**